pandas/_libs/hashtable_class_helper.pxi.in

```diff
@@ -243,17 +243,24 @@ cdef class {{name}}Vector(Vector):
 
     cdef void append(self, {{c_type}} x) noexcept:
 
-        if needs_resize(self.data):
+        if needs_resize(self.data.n, self.data.m):
             if self.external_view_exists:
                 raise ValueError("external reference but "
                                  "Vector.resize() needed")
-            self.resize()
+            self.resize(self.data.m * 4)
 
         append_data_{{dtype}}(self.data, x)
 
     cdef extend(self, const {{c_type}}[:] x):
-        for i in range(len(x)):
-            self.append(x[i])
+        x_size = len(x)
+        new_size = self.data.n + x_size
+        if needs_resize(new_size, self.data.m):
+            if self.external_view_exists:
+                raise ValueError("external reference but "
+                                 "Vector.resize() needed")
+            self.resize(new_size)  # TODO: do we want to multiply by 4?
```

Comment:
With the `extend` I choose not to multiple by a factor of 4. This would favor performance when doing really large appends versus many small appends in a loop 

Not sure how to best manage this. Long term I would love to get rid of this self-managed code and use the C++ standard library